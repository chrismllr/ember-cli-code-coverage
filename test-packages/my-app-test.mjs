'use strict';

import fs from 'fs-extra';
import { execa } from 'execa';
import setupTestDir, { assertCoverageExists, assertDirDoesNotExists } from './utils.mjs';
import { describe, it } from 'vitest';

const APP_DIR = 'my-app';
const APP_PATH = `test-packages/${APP_DIR}`;

describe('app coverage generation', function () {
  const env = { COVERAGE: 'true' };

  it('runs coverage when env var is set', async function () {
    let buildPath = await setupTestDir(APP_DIR, env, {});

    await execa('npx', ['ember', 'test', '--test-port=0'], { cwd: buildPath, env });

    await assertCoverageExists(`${buildPath}/coverage`);
  });

  it('does not run coverage when env var is NOT set', async function () {
    let env = { COVERAGE: 'false' };
    let buildPath = await setupTestDir(APP_DIR, env, {});

    await execa('npx', ['ember', 'test', '--test-port=0'], { cwd: buildPath, env });

    await assertDirDoesNotExists(`${buildPath}/coverage`);
  });

  it('excludes files when the configuration is set', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});
    fs.copySync(`${APP_PATH}/config/-coverage-excludes.js`, `${buildPath}/config/coverage.js`);

    await execa('npx', ['ember', 'test', '--test-port=0'], { cwd: buildPath, env });

    await assertCoverageExists(`${buildPath}/coverage`);
  });

  it('runs coverage when the path option is used', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});

    await execa('pnpm', ['ember', 'build', '--output-path=test-dist'], { cwd: buildPath, env });
    await execa('pnpm', ['ember', 'test', '--path=test-dist', '--test-port=0'], { cwd: buildPath, env });

    await assertCoverageExists(`${buildPath}/coverage`);
  });

  it('merges coverage when tests are run in parallel', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});
    await execa('pnpm', ['ember', 'exam', '--split=2', '--parallel=true'], { cwd: buildPath, env });
    await assertCoverageExists(`${buildPath}/coverage`);
  });

  it('uses parallel configuration and merges coverage when merge-coverage command is issued', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});
    fs.copySync(`${APP_PATH}/config/-coverage-parallel.js`, `${buildPath}/config/coverage.js`);

    await execa('pnpm', ['ember', 'exam', '--split=2', '--parallel=true'], { cwd: buildPath, env });
    await assertDirDoesNotExists(`${buildPath}/coverage`);

    await execa('pnpm', ['ember', 'coverage-merge'], { cwd: buildPath });

    await assertCoverageExists(`${buildPath}/coverage`);
  });

  it('uses nested coverageFolder and parallel configuration and run merge-coverage', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});
    let coverageFolder = `${buildPath}/coverage/abc/easy-as/123`;

    await assertDirDoesNotExists(coverageFolder);
    fs.copySync(`${APP_PATH}/config/-coverage-nested-folder.js`, `${buildPath}/config/coverage.js`);

    await execa('pnpm', ['ember', 'exam', '--split=2', '--parallel=true'], { cwd: buildPath, env });
    await assertDirDoesNotExists(coverageFolder);

    await execa('pnpm', ['ember', 'coverage-merge'], { cwd: buildPath });
    await assertCoverageExists(coverageFolder);
  });

  it('runs coverage when a module has an import error', async function () {
    let env = { COVERAGE: 'true' };
    let buildPath = await setupTestDir(APP_DIR, env, {});
    fs.copySync(`${APP_PATH}/-error-module.js`, `${buildPath}/app/error-module.js`);

    await execa('pnpm', ['ember', 'test',' --test-port=0'], { cwd: buildPath, env });

    await assertCoverageExists(`${buildPath}/coverage`);
  });
});
