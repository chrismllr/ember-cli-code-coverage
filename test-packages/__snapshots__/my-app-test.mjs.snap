// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app coverage generation > excludes files when the configuration is set 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
    "statements": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
  },
}
`;

exports[`app coverage generation > merges coverage when tests are run in parallel 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
    "statements": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
  },
}
`;

exports[`app coverage generation > runs coverage when a module has an import error 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/error-module.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 7,
      "pct": 77.77,
      "skipped": 0,
      "total": 9,
    },
    "statements": {
      "covered": 7,
      "pct": 77.77,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;

exports[`app coverage generation > runs coverage when env var is set 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
    "statements": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
  },
}
`;

exports[`app coverage generation > runs coverage when the path option is used 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
    "statements": {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
  },
}
`;

exports[`app coverage generation > uses nested coverageFolder and parallel configuration and run merge-coverage 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": {
      "covered": 6,
      "pct": 66.66,
      "skipped": 0,
      "total": 9,
    },
    "statements": {
      "covered": 6,
      "pct": 66.66,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;

exports[`app coverage generation > uses parallel configuration and merges coverage when merge-coverage command is issued 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": {
      "covered": 6,
      "pct": 66.66,
      "skipped": 0,
      "total": 9,
    },
    "statements": {
      "covered": 6,
      "pct": 66.66,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;
